feat: Add more uniform reasoner function with options

smessie opened the pull request:
Closes #8

jeswr commented:
Great work @smessie - the general API looks good and will do a more thorough review of the internals in the morning.

Could you please also document the new options in the Readme?

jeswr commented on lib/transformers.ts:
```diff
@@ -32,45 +32,101 @@ export function SwiplEye(options?: Partial<EmscriptenModule> | undefined) {
  * @returns The same SWIPL module
  */
-export function runQuery(Module: SWIPLModule, data: string, queryString?: string): SWIPLModule {
+export function runQuery(Module: SWIPLModule, data: string, queryString?: string, options = { output: 'derivations', blogic: false }): SWIPLModule {
```
Could you properly define the type of the options using an interface; I expect it should look something like

```ts
IQueryOptions {
  blogic?: boolean;
  outputType?: 'string' | 'quads'
  output?: 'derivations' | 'deductive_closure' | 'deductive_closure_plus_rules' | 'grounded_deductive_closure_plus_rules'
}
```

jeswr commented on README.md:
```diff
@@ -99,6 +99,39 @@ const dataQuads = [
+The `options` parameter is optional and can be used to configure the reasoning process. The following options are available:
+- `output`: What to output.
+    - `derivations`: output only new derived triples, a.k.a `--pass-only-new` (default)
+    - `deductive_closure`: output deductive closure, a.k.a `--pass`
+    - `deductive_closure_plus_rules`: output deductive closure plus rules, a.k.a `--pass-all`
+    - `grounded_deductive_closure_plus_rules`: ground the rules and output deductive closure plus rules, a.k.a `--pass-all-ground`
```
For usability I'd be inclined to make this a few boolean options such as

- `onlyNew`: boolean (toggle between derivations and deductive closure)
- `outputRules`: boolean (include rules in the result)
- `groundRules`: boolean (include ground rules in result).

Is there a reason you have chosen to use the current options instead?

smessie replied:
Couldn't that lead to conflicts? I assumed that you can't combine multiple "pass" flags like `--pass-only-new` and `--pass-all` which would be the case if you configure both `onlyNew` and `outputRules` to be true. But maybe my assumption was incorrect?

jeswr replied:
> I assumed that you can't combine multiple "pass" flags like --pass-only-new and --pass-all which would be the case if you configure both onlyNew and outputRules to be true.

@josd can you answer this

---

That said I think it is fine for there just to be runtime errors if users try and do non-sensible options like that - in the same way that I assume you would get a runtime error if you did `--pass-only-new --pass-all` in eye.

josd replied:
> > I assumed that you can't combine multiple "pass" flags like --pass-only-new and --pass-all which would be the case if you configure both onlyNew and outputRules to be true.
> 
> @josd can you answer this

This is seen as multiple queries, just like you can do `--query q1.n3 --query q2.n3` and you will get the answer for each query.

smessie replied:
It doesn't really make sense to me to allow multiple at once. I agree that booleans look like a cleaner solution but that gives the wrong impression that you can combine them in one query, while in fact they are multiple queries as Jos pointed out. 
Also `groundRules` sounds to me that it only ground the rules, but in fact they ground the rules and then output all the rules in the result, now people might think that they need to set both `groundRules` and `outputRules` to true to achieve this goal.
I think I like your suggestion of the `IQueryOptions` interface more. What do you think @jeswr ?

jeswr replied:
Lets go with the `IQueryOptions` for now and see whether they cause any problems for users.

jeswr commented on lib/transformers.ts:
```diff
@@ -32,45 +32,101 @@ export function SwiplEye(options?: Partial<EmscriptenModule> | undefined) {
+  data: Quad[] | string,
+  query?: Quad[] | string,
+  options = { output: 'derivations', blogic: false, outputType: 'quads' },
+): Promise<Quad[] | string> {
```
Could you add a type overload so that the return type is wither `Promise<Quad[]>` or `Promise<string>` depending on how `outputType` is set (importantly this will allow us to make this a mver bump rather than a breaking change)? Let me know if you need help with this one, type overloads can be finnicky.

smessie replied:
So you mean to have `function executeBasicEyeQueryQuads(swipl, data: Quad[] | string, query?: Quad[] | string, options): Promise<Quad[]>` and a function `function executeBasicEyeQueryQuads(swipl, data: Quad[] | string, query?: Quad[] | string, options): Promise<String>` based on the parameter `outputType` in `options`? (And idem for `executeBasicEyeQuery`)
Then I could use some help because I can't directly see how to do this based on a value of a parameter. 

Alternatively, it could make sense to have (keep) `function executeBasicEyeQuery(swipl, data: Quad[] | string, query?: Quad[] | string, options): Promise<string>` only allowing `outputType` `string` and `function executeBasicEyeQueryQuads(swipl, data: Quad[] | string, query?: Quad[] | string, options): Promise<Quad[]>` only allowing `outputType` `quads`.

jeswr replied:
@smessie in https://github.com/eyereasoner/eye-js/pull/89/commits/24e04cb808894eba10471082f0186f07999067b8 I've added these type overloads in addition to adding the `IQueryOptions` type. It means that the type system will now infer whether the output is `Promise<Quad[]>` or `Promise<string>` based on the input parameters.

I also made one change to the `basicQuery` and `n3Reasoner` API, rather than always defaulting to an output type of `string` it will return the same type as the input data unless the `outputType` is explicitly defined.

jeswr commented:
@smessie before I merge this could you check that you are ok with the changes I made in https://github.com/eyereasoner/eye-js/pull/89/commits/24e04cb808894eba10471082f0186f07999067b8 and also add some test cases for the `blogic` and `output` parameters on the `n3Reasoner` function.

jeswr commented on README.md:
```diff
@@ -99,6 +99,39 @@ const dataQuads = [
+:Alice a :Person .
+{ ?S a :Person } => { ?S a :Human } .
+`;
+
+const query = `{?S ?P ?O . } => {?S ?P ?O . } .`;
+
+const options = { output: "derivations", blogic: false, outputType: 'string' };
+
+const result = await n3reasoner(data, query, options);
```
```suggestion

const result = await n3reasoner(data, undefined, { output: 'derivations', blogic: false, outputType: 'string' });
```

smessie commented on lib/transformers.ts:
```diff
@@ -37,29 +43,26 @@ export function SwiplEye(options?: Partial<EmscriptenModule> | undefined) {
-  if (unknownOptions.length > 0) {
-    throw new Error(
-      `Unknown options: ${unknownOptions.join(', ')}`,
-    );
```
Would you remove this as well? Do you prefer silently ignoring extra unknown options rather than failing on them?

jeswr replied:
Typescript consumers will get a build error if they try and use keys outside those that are defined in the `IQueryOptions` interface; and most editors will give type-hints to JS users as well.

I don't think having runtime errors is necessary in addition to this.

![image](https://user-images.githubusercontent.com/63333554/216765059-15941fbc-3a79-4648-848e-8c751db59b6b.png)

smessie replied:
Fair point! Adding the interface makes this obsolete.

jeswr commented on tsconfig.json:
```diff
@@ -15,7 +15,7 @@
     // "outFile": "./",                       /* Concatenate and emit output to single file. */
     "outDir": "./dist",                        /* Redirect output structure to the directory. */
-    "rootDir": "./lib",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
+    "rootDirs": ["./lib", "./data"],          /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
```
Why was this change made?

smessie replied:
To resolve following error on `import { data, query, result } from '../data/socrates';` in `util.ts`
```
TS6059: File '/Users/ieben/WebstormProjects/eyereasoner/data/socrates.ts' is not under 'rootDir' '/Users/ieben/WebstormProjects/eyereasoner/lib'. 'rootDir' is expected to contain all source files.
```

jeswr replied:
I suspect that you were trying to use one of those values in the core lib as well as the `rootDir` should only contain those files we wish to include in the build.

I have reverted the config and not had any build problems.

jeswr commented on __tests__/util.ts:
```diff
@@ -1,8 +1,9 @@
 /* eslint-disable default-case */
+import { Quad } from '@rdfjs/types';
```
```suggestion
import type { Quad } from '@rdfjs/types';
```

jeswr commented on .eslintrc.json:
```diff
@@ -20,6 +20,10 @@
       "import/extensions": "off",
-      "no-undef": "off"
+      "no-undef": "off",
+      "no-unused-vars": "off",
```
I believe (and possibly some of the other rules) can be re-enabled after [this is fixed](https://github.com/eyereasoner/eye-js/pull/89/files#r1096525713)

smessie replied:
It is actually to fix ESLint errors in the code itself, not in the tests, stuff like all this
<img width="720" alt="image" src="https://user-images.githubusercontent.com/13602120/216768194-52c48e5f-02b9-4164-a813-2f45294a8ac5.png">
I checked with the linked change applied and it still gives those errors.

smessie commented on __tests__/util.ts:
```diff
@@ -22,10 +23,36 @@ export function mockFetch(...args: Parameters<typeof fetch>): ReturnType<typeof
+    it('should execute the basicQuery [quad input explicit string output]', async () => {
+      const resultStr: string = await basicQuery(dataQuads, queryQuads, { outputType: 'string' });
+      const quads = (new Parser({ format: 'text/n3' })).parse(resultStr);
+      expect<Quad[]>(quads).toBeRdfIsomorphic(quads);
+    });
+
+    it('should execute the basicQuery [string input string output]', async () => {
+      const resultStr: string = await basicQuery(data, query);
+      const quads = (new Parser({ format: 'text/n3' })).parse(resultStr);
+      expect<Quad[]>(quads).toBeRdfIsomorphic(quads);
+    });
+
+    it('should execute the basicQuery [string input explicit string output]', async () => {
+      const resultStr: string = await basicQuery(data, query, { outputType: 'string' });
+      const quads = (new Parser({ format: 'text/n3' })).parse(resultStr);
+      expect<Quad[]>(quads).toBeRdfIsomorphic(quads);
```
```suggestion
      expect<Quad[]>(quads).toBeRdfIsomorphic(resultQuads);
    });

    it('should execute the basicQuery [string input string output]', async () => {
      const resultStr: string = await basicQuery(data, query);
      const quads = (new Parser({ format: 'text/n3' })).parse(resultStr);
      expect<Quad[]>(quads).toBeRdfIsomorphic(resultQuads);
    });

    it('should execute the basicQuery [string input explicit string output]', async () => {
      const resultStr: string = await basicQuery(data, query, { outputType: 'string' });
      const quads = (new Parser({ format: 'text/n3' })).parse(resultStr);
      expect<Quad[]>(quads).toBeRdfIsomorphic(resultQuads);
```

jeswr replied:
Good catch @smessie ! Merge this one when you're ready (I don't want to create conflicts if you're doing anything locally).

smessie commented:
@jeswr I've added some tests for the `blogic` and `output` options. Additionally, I've also added a `none` output option and made that the default as well to better reflect the possibilities in eye after discussion with @josd, as no pass is actually the default. 
This also yields the expected output for the blogic option as this expects no -- both explicit and implicit -- query. Without `none` there was no possibility to have no (explicit and implicit) query.
This also means that doing `basicQuery(dataQuads)` yields no input (because default `output` option is `none` now), but that sounds reasonable to me, as you can't expect an answer if you don't ask a question (query).

jeswr commented:
Thanks for being so quick with those additions @smessie - 2 questions:
1. When I run the test suite I now get the warning

```
   console.warn
      ** ERROR ** file:/// ** error(io_error(read,<stream>(0x759470)),context(system:get_code/2,Is a directory))

      578 |                     tty.output = [];
      579 |                 } } }, default_tty1_ops: { put_char: function (tty, val) { if (val === null || val === 10) {
    > 580 |                     err(UTF8ArrayToString(tty.output, 0));
          |                     ^
      581 |                     tty.output = [];
      582 |                 }
      583 |                 else {
```
do you know when that was introduce

2. Rather than using the string `"none"` as the default can say that is the behavior when the `output` field is left undefined; or `undefined` is explicitly supplied as the input.

jeswr commented:
@smessie thanks for all your work, and thankyou @josd for the input provided as well.

I believe this is ready to merge; can I get confirmation that you are happy with the state of this PR before I do so?

smessie commented on README.md:
````diff
@@ -9,97 +9,50 @@ Distributing the [EYE](https://github.com/eyereasoner/eye) reasoner for browser
 ```
 
-Here the inputs and outputs are both arrays of RDF/JS Quads; for instance
+The `n3Reasoner` accepts both `string`s (formatted in Notation3 syntax) and `quad`s as input. The output will be of the same type as the input `data`. This means that we can use `n3Reasoner` with RDF/JS quads as follows:
````
```suggestion
The `n3reasoner` accepts both `string`s (formatted in Notation3 syntax) and `quad`s as input. The output will be of the same type as the input `data`. This means that we can use `n3reasoner` with RDF/JS quads as follows:
```

smessie commented on README.md:
```diff
@@ -9,97 +9,50 @@ Distributing the [EYE](https://github.com/eyereasoner/eye) reasoner for browser
-The simplest way to use this package is to execute a query over a dataset and get the results
-
-**Note**: the `dataQuads` should include any inference rules that you wish to apply to the dataset
+The simplest way to use this package is to use the `n3Reasoner` to execute a query over a dataset and get the results. The input `data` should include the data and any inference rules that you wish to apply to the dataset; the optional `query` should match the pattern of data you wish the engine to return; if left undefined, all inferred facts will be returned. For example:
```
```suggestion
The simplest way to use this package is to use the `n3reasoner` to execute a query over a dataset and get the results. The input `data` should include the data and any inference rules that you wish to apply to the dataset; the optional `query` should match the pattern of data you wish the engine to return; if left undefined, all inferred facts will be returned. For example:
```

smessie commented:
> @smessie thanks for all your work, and thankyou @josd for the input provided as well.

Likewise!

> I believe this is ready to squash & merge; can I get confirmation that you are happy with the state of this PR before I do so?

Looks good to me!

jeswr commented:
:tada: This PR is included in version 2.10.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/eyereasoner/v/2.10.0)
- [GitHub release](https://github.com/eyereasoner/eye-js/releases/tag/v2.10.0)
- `v2.10.0`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
